File: src/main/scala/perfio/VectorizedLineTokenizer.scala
package perfio

import jdk.incubator.vector.{ByteVector, VectorMask, VectorOperators}

import java.lang.{Long => JLong}
import java.nio.ByteOrder
import java.nio.charset.{Charset, StandardCharsets}
import scala.annotation.tailrec

/**
 * A vectorized implementation of [[LineTokenizer]].
 *
 * The parser uses explicit SIMD loops with up to 512 bits / 64 lanes per vector (depending on
 * hardware and JVM support).
 */
object VectorizedLineTokenizer {

  /** Create a VectorizedLineTokenizer2 from a BufferedInput. See [[LineTokenizer.apply]] for details. */
  def apply(in: BufferedInput, charset: Charset = StandardCharsets.UTF_8, eol: Byte = '\n'.toByte, preEol: Byte = '\r'.toByte): LineTokenizer = in match {
    case in: HeapBufferedInput =>
      if(charset eq StandardCharsets.ISO_8859_1) new HeapVectorizedLineTokenizer(in, eol, preEol) {
        protected[this] def makeString(buf: Array[Byte], start: Int, len: Int): String = new String(buf, 0, start, len)
      } else new HeapVectorizedLineTokenizer(in, eol, preEol) {
        protected[this] def makeString(buf: Array[Byte], start: Int, len: Int): String = new String(buf, start, len, charset)
      }
    case in: DirectBufferedInput =>
      if(charset eq StandardCharsets.ISO_8859_1) new DirectVectorizedLineTokenizer(in, eol, preEol) {
        protected[this] def makeString(buf: Array[Byte], start: Int, len: Int): String = new String(buf, 0, start, len)
      } else new DirectVectorizedLineTokenizer(in, eol, preEol) {
        protected[this] def makeString(buf: Array[Byte], start: Int, len: Int): String = new String(buf, start, len, charset)
      }
  }
}

private sealed abstract class HeapVectorizedLineTokenizer(_bin: HeapBufferedInput, _eolChar: Byte, _preEolChar: Byte) extends
  HeapLineTokenizer(_bin, _eolChar, _preEolChar) {
  import VectorSupport._

  private[this] var vpos = bin.pos-vlen // start of the current vector in buf
  private[this] var mask = 0L // vector mask that marks the LFs
  private[this] val eolVector = ByteVector.broadcast(species, eolChar)

  if(bin.pos < bin.lim) {
    // Make sure the position is aligned if we have buffered data.
    // Otherwise we leave the initial buffering to the main loop in readLine().
    vpos += vlen
    alignPosAndComputeMask()
  }

  private[this] def buffer(): Int = {
    //println(s"buffer: $closed, $mask, $vpos")
    //println(s"  buffer(): pos=${bin.pos}, lim=${bin.lim}, vpos=$vpos, buflen=${bin.buf.length}")
    if(vpos == Int.MaxValue) {
      vpos = Int.MaxValue - vlen
      return 0
    }
    val scanned = bin.lim - bin.pos
    val oldav = bin.available
    bin.prepareAndFillBuffer(oldav + vlen)
    //println(s"    $oldav -> ${bin.available}")
    if(oldav == bin.available) return 0
    vpos = bin.pos + scanned
    alignPosAndComputeMask()
    1
  }

  override private[perfio] def markClosed(): Unit = {
    super.markClosed()
    mask = 0L
    vpos = Int.MaxValue - vlen
  }

  private[this] def alignPosAndComputeMask(): Unit = {
    val offsetInVector = vpos % vlen
    vpos -= offsetInVector
    mask = computeMask() & (-1L >>> offsetInVector << offsetInVector)
  }

  private[this] def rest(): String = {
    checkState()
    //println(s"rest: pos=$pos")
    val r = if(vpos != Int.MaxValue) {
      val l = bin.available
      mask = 0
      if(l != 0) {
        val s = makeString(bin.buf, bin.pos, l)
        bin.pos = bin.lim
        s
      } else null
    } else null
    vpos = Int.MaxValue - vlen
    r
  }

  @tailrec final def readLine(): String = {
    val f = JLong.numberOfTrailingZeros(mask)
    //println(s"readLine(): ${bin.pos}, ${bin.lim}, $vpos, $mask -> $f")
    if(f < vlen) {
      val lfpos = vpos+f
      val s = emit(bin.pos, lfpos)
      bin.pos = lfpos+1
      mask &= ~(1L<<f)
      return s
    }
    vpos += vlen
    if(vpos <= bin.lim-vlen) mask = computeSimpleMask()
    else if(vpos < bin.lim) mask = computeMask()
    else if(buffer() == 0) return rest()
    readLine()
  }

  private[this] final def computeSimpleMask(): Long =
    ByteVector.fromArray(species, bin.buf, vpos).compare(VectorOperators.EQ, eolVector).toLong

  private[this] final def computeMask(): Long = {
    val excess = vpos+vlen-bin.lim
    val v =
      if(excess <= 0) ByteVector.fromArray(species, bin.buf, vpos)
      else ByteVector.fromArray(species, bin.buf, vpos, VectorMask.fromLong(species, fullMask >>> excess))
    // Seems to be slightly faster than comparing against a scalar:
    v.compare(VectorOperators.EQ, eolVector).toLong
  }
}

private sealed abstract class DirectVectorizedLineTokenizer(_bin: DirectBufferedInput, _eolChar: Byte, _preEolChar: Byte)
  extends DirectLineTokenizer(_bin, _eolChar, _preEolChar) {
  import VectorSupport._

  private[this] var vpos = start-vlen
  private[this] val limit = ms.byteSize()
  private[this] var mask = 0L
  private[this] val eolVector = ByteVector.broadcast(species, eolChar)

  override private[perfio] def markClosed(): Unit = {
    super.markClosed()
    mask = 0L
    vpos = Long.MaxValue - vlen
  }

  private[this] def rest(): String = {
    val r = if(vpos != Long.MaxValue) {
      val l = limit-start
      mask = 0
      if(l != 0) {
        val s = makeString(ms, start, l)
        start = limit
        s
      } else null
    } else {
      checkState()
      null
    }
    vpos = Long.MaxValue - vlen
    r
  }

  @tailrec final def readLine(): String = {
    val f = JLong.numberOfTrailingZeros(mask)
    if(f < vlen) {
      val lfpos = vpos+f
      val s = emit(start, lfpos)
      start = lfpos+1
      mask &= ~(1L<<f)
      return s
    }
    vpos += vlen
    if(vpos <= limit-vlen) mask = computeSimpleMask()
    else if(vpos < limit) mask = computeMask()
    else return rest()
    readLine()
  }

  @inline private[this] final def computeSimpleMask(): Long =
    ByteVector.fromMemorySegment(species, ms, vpos, ByteOrder.BIG_ENDIAN).compare(VectorOperators.EQ, eolVector).toLong

  private[this] final def computeMask(): Long = {
    val excess = vpos+vlen-limit
    val v =
      if(excess <= 0) ByteVector.fromMemorySegment(species, ms, vpos, ByteOrder.BIG_ENDIAN)
      else ByteVector.fromMemorySegment(species, ms, vpos, ByteOrder.BIG_ENDIAN, VectorMask.fromLong(species, fullMask >>> excess))
    // Seems to be slightly faster than comparing against a scalar:
    v.compare(VectorOperators.EQ, eolVector).toLong
  }
}

File: src/main/java/perfio/VectorizedLineTokenizer.java
package perfio;

import jdk.incubator.vector.ByteVector;
import jdk.incubator.vector.VectorMask;
import jdk.incubator.vector.VectorOperators;

import java.io.IOException;
import java.lang.foreign.MemorySegment;
import java.nio.ByteOrder;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import static perfio.VectorSupport.*;

/// A vectorized implementation of [[LineTokenizer]].
///
/// The parser uses explicit SIMD loops with up to 512 bits / 64 lanes per vector (depending on
/// hardware and JVM support).
public class VectorizedLineTokenizer {
  private VectorizedLineTokenizer() {}

  static LineTokenizer of(BufferedInput in, Charset charset, byte eol, byte preEol) throws IOException {
    if(in instanceof HeapBufferedInput h) {
      if(charset == StandardCharsets.ISO_8859_1) return new HeapVectorizedLineTokenizer(h, eol, preEol) {
        String makeString(byte[] buf, int start, int len) { return new String(buf, 0, start, len); }
      };
      else return new HeapVectorizedLineTokenizer(h, eol, preEol) {
        String makeString(byte[] buf, int start, int len) { return new String(buf, start, len, charset); }
      };
    } else {
      var i = (DirectBufferedInput)in;
      if(charset == StandardCharsets.ISO_8859_1) {
        if(StringInternals.internalAccessEnabled) return new DirectVectorizedLineTokenizer(i, eol, preEol) {
          String makeString(byte[] buf, int start, int len) { return new String(buf, 0, start, len); }
          @Override String makeString(MemorySegment buf, long start, long llen) { return makeStringLatin1Internal(buf, start, llen); }
        };
        else return new DirectVectorizedLineTokenizer(i, eol, preEol) {
          String makeString(byte[] buf, int start, int len) { return new String(buf, 0, start, len); }
        };
      } else return new DirectVectorizedLineTokenizer(i, eol, preEol) {
        String makeString(byte[] buf, int start, int len) { return new String(buf, start, len, charset); }
      };
    }
  }

  static LineTokenizer of(BufferedInput in, Charset charset) throws IOException { return of(in, charset, (byte)'\n', (byte)'\r'); }
}


abstract non-sealed class HeapVectorizedLineTokenizer extends HeapLineTokenizer {
  private int vpos; // start of the current vector in buf
  private long mask = 0L; // vector mask that marks the LFs
  private final ByteVector eolVector;

  HeapVectorizedLineTokenizer(HeapBufferedInput parentBin, byte eolChar, byte preEolChar) throws IOException {
    super(parentBin, eolChar, preEolChar);
    this.vpos = bin.pos - VLEN;
    this.mask = 0L;
    this.eolVector = ByteVector.broadcast(SPECIES, eolChar);
    if(bin.pos < bin.lim) {
      // Make sure the position is aligned if we have buffered data.
      // Otherwise we leave the initial buffering to the main loop in readLine().
      vpos += VLEN;
      alignPosAndComputeMask();
    }
  }

  private int buffer() throws IOException {
    if(vpos == Integer.MAX_VALUE) {
      vpos = Integer.MAX_VALUE - VLEN;
      return 0;
    }
    var scanned = bin.lim - bin.pos;
    var oldav = bin.available();
    bin.prepareAndFillBuffer(oldav + VLEN);
    if(oldav == bin.available()) return 0;
    vpos = bin.pos + scanned;
    alignPosAndComputeMask();
    return 1;
  }

  @Override
  public void markClosed() {
    super.markClosed();
    mask = 0L;
    vpos = Integer.MAX_VALUE - VLEN;
  }

  private void alignPosAndComputeMask() {
    var offsetInVector = vpos % VLEN;
    vpos -= offsetInVector;
    mask = computeMask() & (-1L >>> offsetInVector << offsetInVector);
  }

  private String rest() throws IOException {
    checkState();
    String r;
    if(vpos != Integer.MAX_VALUE) {
      var l = bin.available();
      mask = 0;
      if(l != 0) {
        var s = makeString(bin.buf, bin.pos, l);
        bin.pos = bin.lim;
        r = s;
      } else r = null;
    } else r = null;
    vpos = Integer.MAX_VALUE - VLEN;
    return r;
  }

  public final String readLine() throws IOException {
    while(true) {
      var f = Long.numberOfTrailingZeros(mask);
      //println(s"readLine(): ${bin.pos}, ${bin.lim}, $vpos, $mask -> $f")
      if(f < VLEN) {
        var lfpos = vpos+f;
        var s = emit(bin.pos, lfpos);
        bin.pos = lfpos+1;
        mask &= ~(1L<<f);
        return s;
      }
      vpos += VLEN;
      if(vpos <= bin.lim-VLEN) mask = computeSimpleMask();
      else if(vpos < bin.lim) mask = computeMask();
      else if(buffer() == 0) return rest();
    }
  }
  private long computeSimpleMask() {
    return ByteVector.fromArray(SPECIES, bin.buf, vpos).compare(VectorOperators.EQ, eolVector).toLong();
  }

  private long computeMask() {
    var excess = vpos+VLEN-bin.lim;
    var v = excess <= 0
        ? ByteVector.fromArray(SPECIES, bin.buf, vpos)
        : ByteVector.fromArray(SPECIES, bin.buf, vpos, VectorMask.fromLong(SPECIES, FULL_MASK >>> excess));
    // Seems to be slightly faster than comparing against a scalar:
    return v.compare(VectorOperators.EQ, eolVector).toLong();
  }
}


abstract non-sealed class DirectVectorizedLineTokenizer extends DirectLineTokenizer {
  private long vpos;
  private final long limit;
  private long mask = 0L;
  private final ByteVector eolVector;

  DirectVectorizedLineTokenizer(DirectBufferedInput bin, byte eolChar, byte preEolChar) throws IOException {
    super(bin, eolChar, preEolChar);
    this.vpos = start - VLEN;
    this.limit = ms.byteSize();
    this.eolVector = ByteVector.broadcast(SPECIES, eolChar);
  }

  @Override
  public void markClosed() {
    super.markClosed();
    mask = 0L;
    vpos = Long.MAX_VALUE - VLEN;
  }

  private String rest() throws IOException {
    String r;
    if(vpos != Long.MAX_VALUE) {
      var l = limit-start;
      mask = 0;
      if(l != 0) {
        var s = makeString(ms, start, l);
        start = limit;
        r = s;
      } else r = null;
    } else {
      checkState();
      r = null;
    }
    vpos = Long.MAX_VALUE - VLEN;
    return r;
  }

  public String readLine() throws IOException {
    while(true) {
      var f = Long.numberOfTrailingZeros(mask);
      if(f < VLEN) {
        var lfpos = vpos+f;
        var s = emit(start, lfpos);
        start = lfpos+1;
        mask &= ~(1L<<f);
        return s;
      }
      vpos += VLEN;
      if(vpos <= limit-VLEN) mask = computeSimpleMask();
      else if(vpos < limit) mask = computeMask();
      else return rest();
    }
  }

  private long computeSimpleMask() {
    return ByteVector.fromMemorySegment(SPECIES, ms, vpos, ByteOrder.BIG_ENDIAN).compare(VectorOperators.EQ, eolVector).toLong();
  }

  private long computeMask() {
    var excess = vpos+VLEN-limit;
    var v = excess <= 0
        ? ByteVector.fromMemorySegment(SPECIES, ms, vpos, ByteOrder.BIG_ENDIAN)
        : ByteVector.fromMemorySegment(SPECIES, ms, vpos, ByteOrder.BIG_ENDIAN, VectorMask.fromLong(SPECIES, FULL_MASK >>> excess));
    // Seems to be slightly faster than comparing against a scalar:
    return v.compare(VectorOperators.EQ, eolVector).toLong();
  }
}
